syntax="proto3";

message Point {
  int32 x = 1;
  int32 y = 2;
}

message Snake {
  string color = 1;
  string direction = 2;
  repeated Point body = 3;
  bool isBot = 4;
  string name = 5;
  int32 score = 6;
}

message JoinRequest {
  int32 windowWidth= 1;
  int32 windowHeight= 2;
  int32 snakeSize = 3;
}

message MoveRequest {
  string color = 1;
  string direction = 2;
}

message CollisionRequest{
  string color = 1;
}
message CollisionResponse {
  bool has_collided = 1;
}

message GetRequest {}
message FoodRequest {}

message BoardDimension {
  int32 width = 1;
  int32 height = 2;
}

service SnakeService {
  rpc getBoardDimension (JoinRequest) returns (BoardDimension) {}
  rpc addSnake (GetRequest) returns (Snake) {}
  rpc removeSnake (Snake) returns (Snake) {}
  rpc moveSnake (MoveRequest) returns (Snake) {}
  rpc getSnakes (GetRequest) returns (stream Snake){}
  rpc getFood (FoodRequest) returns (stream Point) {}
  rpc addMoreFood (FoodRequest) returns (FoodRequest) {}
  rpc checkCollision(CollisionRequest) returns (CollisionResponse){}
}